### PR TITLE
ARROW-16131 [C++] support saving and retrieving custom metadata in batches for IPC file

### PR DESCRIPTION
**Pull request by niyue**
This PR aims to address https://issues.apache.org/jira/projects/ARROW/issues/ARROW-16131

Currently, when writing an IPC file with multiple record batches using the `arrow::ipc::RecordBatchWriter`, the `custom_metadata` for each record batch is not saved and will be discarded silently. 

The current writer does provide the `AppendCustomMetadata` API internally, but it is never called. I use this API to add the custom metadata during writing and retrieve the custom metadata when reading the record batch. I am not completely sure if this is intentionally ignored or accidentally missing, but from a user's perspective, the metadata can be set via public API (see the example case in ARROW-16131) and I think it is not very friendly that this piece of data is silently discarded.

Several test cases are added in the `read_write_test.cc` for verifying this change.

### CODE REVIEW

**Comment by niyue** [comment]
I checked out all the build failures reported , including `arrow-gcsfs-test (Timeout)` and `arrow-s3-test (Failed)`, and they don't seem to be related with my change. 

I can build and run the S3 test locally, and it ran into segfault locally as the CI server did for the test case `GetFileInfoRoot`:
```
TEST_F(TestS3FS, GetFileInfoRoot) { AssertFileInfo(fs_.get(), "", FileType::Directory); }
```

For the gcs test, I failed to build the google cloud cpp lib and I am not able to run it locally so far, but it seems happened in other PR as well (for example, https://github.com/apache/arrow/pull/12816)

File: cpp/src/arrow/ipc/test_common.cc
Comment (niyue) [inline cpp/src/arrow/ipc/test_common.cc:279]: I add a new test utility function to generate a random int batch with the provided metadata key value pairs.

File: cpp/src/arrow/ipc/writer.cc
Comment (niyue) [inline cpp/src/arrow/ipc/writer.cc:272]: When writing a batch, its batch specific metadata will be copied into the serializer's internal metadata variable. Previously, we have the `AppendCustomMetadata` API, but it is never called, so the serializer's internal metadata variable is always empty.

File: cpp/src/arrow/ipc/reader.cc
Comment (niyue) [inline cpp/src/arrow/ipc/reader.cc:686]: When reading a batch, if the batch message contains some custom metadata, it will be used to create a new schema with the custom metadata when creating the record batch. If the batch message doesn't contain custom metadata, we still use the original schema for creating the record batch as usual.

File: cpp/src/arrow/ipc/writer.cc
Comment (lidavidm) [inline cpp/src/arrow/ipc/writer.cc:267]:
```suggestion
    const auto& metadata = batch.schema()->metadata();
```
for consistency

Comment (lidavidm) [reply cpp/src/arrow/ipc/writer.cc:267]: I'm not sure about this change. If the overall schema has metadata, then won't this copy the schema metadata into every batch, duplicating it over and over?

Comment (lidavidm) [reply cpp/src/arrow/ipc/writer.cc:267]: Maybe we should instead add overloads that also allow explicitly passing in metadata to include with the message. Thoughts @pitrou @westonpace?

Comment (westonpace) [reply cpp/src/arrow/ipc/writer.cc:267]: I think I agree.  This would make it too easy to write an inefficient file that contains a lot of duplicate metadata.  Adding an optional metadata field to the `WriteTable` method makes sense.

Comment (niyue) [reply cpp/src/arrow/ipc/writer.cc:267]: I think there are two cases.

1) in one case, a user simply re-uses the overall schema that happens to have some metadata, and it is less likely this user wants to duplicate the schema metadata in all batches and discard them silently is okay (and duplicating it all record batches may be undesirable and problematic):
```python
    schema = pa.schema(
        [
            ("values", pa.int64()),
        ],
        metadata={"foo": "bar"},
    )
    writer = pa.RecordBatchFileWriter(
        ipc_file, schema
    )
    for i in range(num_batches):
        batch = pa.record_batch(
            [int_array],
            schema=schema # <=== re-use the overall schema having metadata
        )
        writer.write_batch(batch)
```

2) in the other case, a user provides metadata explicitly when creating a record batch, and discarding the metadata silently may not be desirable in this case:
```python
   schema = pa.schema(
        [
            ("values", pa.int64()),
        ],
        metadata={"foo": "bar"},
    )
    writer = pa.RecordBatchFileWriter(
        ipc_file, schema
    )
    for i in range(num_batches):
        batch = pa.record_batch(
            [int_array],
            names=["values"],
            metadata={"batch_id": str(i)}, # <=== pass a metadata explicitly
        )
        writer.write_batch(batch)
```
The underlying implementation is difficult to tell these two cases.

Some current API, like `pyarrow.record_batch`, already allows users to specify metadata explicitly. If we provide an overloaded function asking users to provide metadata, it seems we make the `metadata` in these APIs useless (or only useful when the record batch is in memory). The code may look like:
```python
batch = pa.record_batch(
    [int_array],
    names=["values"] # <== `metadata` passed here will never be used by `write_batch` so users won't pass this parameter any more
)
writer.write_batch(batch, metadata={"batch_id": str(i)})
```

Maybe a boolean flag indicating if the metadata in record batch should be serialized?
```python
batch = pa.record_batch(
    [int_array],
    names=["values"],
    metadata={"batch_id": str(i)}
)
writer.write_batch(batch, serializing_metadata=True)
```

Comment (westonpace) [reply cpp/src/arrow/ipc/writer.cc]: A boolean parameter seems like a fine idea to me.  Let's call it `serialize_metadata` to keep the tenses consistent.

Comment (niyue) [reply cpp/src/arrow/ipc/writer.cc]: @westonpace 
>  Let's call it serialize_metadata to keep the tenses consistent
Sure. I am not a native English speaker, and feel free to let me know if there is better naming for this parameter.

@westonpace @lidavidm 
I will give it a try but I am not completely sure what kind of API compatibility is required. Could you give me some suggestions on what the overloaded function signatures are desirable?

Currently in the C++ API, we have:
```C++
class ARROW_EXPORT RecordBatchWriter {
 public:
  virtual ~RecordBatchWriter();

  /// \brief Write a record batch to the stream
  ///
  /// \param[in] batch the record batch to write to the stream
  /// \return Status
  virtual Status WriteRecordBatch(const RecordBatch& batch) = 0;

  /// \brief Write possibly-chunked table by creating sequence of record batches
  /// \param[in] table table to write
  /// \return Status
  Status WriteTable(const Table& table);

  /// \brief Write Table with a particular chunksize
  /// \param[in] table table to write
  /// \param[in] max_chunksize maximum length of table chunks. To indicate
  /// that no maximum should be enforced, pass -1.
  /// \return Status
  virtual Status WriteTable(const Table& table, int64_t max_chunksize);
```

Do we need to keep all existing APIs unchanged or is it allowed to introduce some compatible API changes?
Some options I can think of:
1) option 1, keep all existing APIs unchanged and add overloaded APIs only
```C++
virtual Status WriteRecordBatch(const RecordBatch& batch) = 0;
virtual Status WriteRecordBatch(const RecordBatch& batch, bool serialize_metadata); // new overload API
Status WriteTable(const Table& table);
virtual Status WriteTable(const Table& table, int64_t max_chunksize);
virtual Status WriteTable(const Table& table, int64_t max_chunksize, bool serialize_metadata); // new overload API
```

2) option 2, add a default parameter to existing APIs
```C++
virtual Status WriteRecordBatch(const RecordBatch& batch, bool serialize_metadatda=false) = 0; // this will require modification to all sub classes
Status WriteTable(const Table& table);
virtual Status WriteTable(const Table& table, int64_t max_chunksize, bool serialize_metadata=false);
```

I assume we expect option 1, is it correct?

Additionally, there may be some higher level binding libraries (Python/Ruby/etc) using these APIs, and they may have to be changed to take advantage of these new APIs, what is our strategy for managing them for such a change? Do I need to change them all in this PR (I think I can try changing Python at least but I am not sure if I am familiar with all the binding languages to make such a change) or should I separate the changes into several PRs? Thanks.

Comment (lidavidm) [reply cpp/src/arrow/ipc/writer.cc]: For the bindings: we can tackle them separately. CI will catch us if we make any breaking changes that need to be handled here. It's only Python/R/GLib (Ruby) we have to worry about (Java/Go have a few bindings I think but not at this level). 

For the APIs: I would prefer adding overloads. CC @pitrou for opinions too.

**Comment by pitrou** [comment]
Hmm, I don't think this is the right way to do this. The IPC message's `custom_metadata` is independent from record batch (i.e. schema) metadata, one should not influence the other.

(also, you could have a `custom_metadata` for other message types such as Tensor)

cc @wesm for opinions.

**Comment by pitrou** [comment]
So, basically, per-message metadata is an Arrow *IPC* feature that's independent from record batches or schemas. That's why you don't see it mentioned in the [schema metadata documentation](https://arrow.apache.org/docs/python/data.html#custom-schema-and-field-metadata).

> If message metadata is different from schema metadata, the question here may be slightly more complex since in this case, each record batch may provide 1) this record batch's schema specific metadata, this seems not serialized/deserialized 2) this record batch's message specific metadata, we don't seem to provide API for reading/writing this

Indeed the message metadata wouldn't be attached to the record batch or schema, but passed separately.

One possible API on the write side:
```c++
class ARROW_EXPORT RecordBatchWriter {
 public:
  // (default implementation could call `WriteRecordBatch(batch)` while ignoring metadata?)
  virtual Status WriteRecordBatch(const RecordBatch& batch, const KeyValueMetadata& custom_metadata);
```

and on the read side:
```c++
class ARROW_EXPORT RecordBatchReader {
 public:
  // (default implementation could call `ReadNext(batch)` and reset `*custom_metadata` to null)
  virtual Status ReadNext(std::shared_ptr<RecordBatch>* batch, std::shared_ptr<KeyValueMetadata>* custom_metadata);
```

**Comment by niyue** [comment]
> One possible API on the write side

Thanks for the advice. Let me give it a try.

File: cpp/src/arrow/ipc/reader.h
Comment (niyue) [inline cpp/src/arrow/ipc/reader.h:200]: This is a new API added for the `RecordBatchFileReader` class so that accessing a record batch with its metadata  by the index is possible.

File: cpp/src/arrow/record_batch.h
Comment (niyue) [inline cpp/src/arrow/record_batch.h:233]: I add this new API with a default implementation so that existing sub class of `RecordBatchReader` can still be compiled successfully.

**Comment by niyue** [comment]
@pitrou @lidavidm 
I submitted a new commit following the suggestion above. I tried to keep the changes compatible with existing public APIs so several new APIs are added. Could you please help to review? Let me know if there is any issue. Thanks.

For the CI:
1) I found one test case `80 - arrow-compute-hash-join-node-test` failed in `C++ / AMD64 Conda C++ (pull_request)`, but I guess this is probably not caused by my PR
2) another test case `51 - arrow-flight-test (Failed)` failed in `continuous-integration/appveyor/pr`, I am not able to find the detailed test report in the CI job, is this an issue?

**Comment by pitrou** [comment]
@niyue Indeed, these CI failures are unrelated. I'll let @lidavidm take a look at the Flight one; the hash-join-node-test crash is at https://issues.apache.org/jira/browse/ARROW-15361

**Comment by lidavidm** [comment]
I filed ARROW-16215

File: cpp/src/arrow/ipc/reader.h
Comment (lidavidm) [inline cpp/src/arrow/ipc/reader.h:193]:
```suggestion
  /// \brief Read a particular record batch along with its custom metadada from the file.
```

File: cpp/src/arrow/record_batch.h
Comment (lidavidm) [reply cpp/src/arrow/record_batch.h:233]: Hmm, should we prefer `arrow::Result` for new APIs?

File: cpp/src/arrow/ipc/reader.cc
Comment (lidavidm) [inline cpp/src/arrow/ipc/reader.cc:909]: you can `std::move(result).ValueUnsafe()`

File: cpp/src/arrow/record_batch.h
Comment (pitrou) [reply cpp/src/arrow/record_batch.h:233]: Perhaps something like:
```c++
struct RecordBatchWithMetadata {
  std::shared_ptr<RecordBatch> batch;
  std::shared_ptr<KeyValueMetadata> custom_metadata;
};
virtual Result<RecordBatchWithMetadata> ReadNext();
```

File: cpp/src/arrow/ipc/reader.h
Comment (pitrou) [reply cpp/src/arrow/ipc/reader.h:200]: I think user code would be more readable with:
```c++
struct RecordBatchWithMetadata {
  std::shared_ptr<RecordBatch> batch;
  std::shared_ptr<KeyValueMetadata> custom_metadata;
};
virtual Result<RecordBatchWithMetadata> ReadRecordBatchWithMetadata(int i) = 0;
```

File: cpp/src/arrow/ipc/reader.h
Comment (niyue) [reply cpp/src/arrow/ipc/reader.h]: Fixed

File: cpp/src/arrow/ipc/reader.cc
Comment (niyue) [reply cpp/src/arrow/ipc/reader.cc]: After changing the API, I switch to use the `ARROW_ASSIGN_OR_RAISE` since it seems more common in the code.

File: cpp/src/arrow/ipc/reader.h
Comment (niyue) [reply cpp/src/arrow/ipc/reader.h]: Sure. I change the code to return the struct now.

File: cpp/src/arrow/record_batch.h
Comment (niyue) [reply cpp/src/arrow/record_batch.h:237]: I change the code to return `Result<RecordBatchWithMetadata>` now.

File: cpp/src/arrow/ipc/reader.cc
Comment (lidavidm) [inline cpp/src/arrow/ipc/reader.cc:867]: nit, but can this be `std::move`d?

**Review by lidavidm — APPROVED** [review]

File: cpp/src/arrow/ipc/reader.cc
Comment (niyue) [reply cpp/src/arrow/ipc/reader.cc]: Fixed.

**Comment by niyue** [comment]
I rebased the PR onto the latest master branch and resolved the conflict in `read_write_test.cc`. 

@pitrou  @lidavidm could you please help to review if there is anything else that needs to be revised? Thanks.

**Review by lidavidm — APPROVED** [review]

**Review by pitrou — APPROVED** [review]
+1, let's wait for CI

**Comment by pitrou** [comment]
Thanks a lot for your contribution @niyue !

**Comment by niyue** [comment]
@pitrou @lidavidm thanks so much for the detailed review

**Pull request closed without merging**
